compositor_render/src/render_loop.rs

```diff
@@ -0,0 +1,58 @@
+use std::{collections::HashMap, sync::Arc};
+
+use compositor_common::{scene::InputId, Frame};
+use log::error;
+
+use crate::renderer::{
+    scene::{Node, Scene},
+    texture::Texture,
+    RenderCtx,
+};
+
+pub(super) fn populate_inputs(
+    ctx: &RenderCtx,
+    scene: &Scene,
+    frames: &mut HashMap<InputId, Arc<Frame>>,
+) {
+    for (input_id, (_, yuv_textures)) in &scene.inputs {
+        Texture::upload_frame_to_textures(
+            ctx.wgpu_ctx,
+            yuv_textures,
+            frames.remove(input_id).unwrap(),
+        );
+        error!("TODO: convert yuv to rgba")
+    }
+}
+
+pub(super) fn run_transforms(ctx: &RenderCtx, scene: &Scene) {
+    for node in graph_iter(scene) {
+        let sources: Vec<_> = node
+            .inputs
+            .iter()
+            .map(|node| (&node.node_id, &node.output))
+            .collect();
+        node.transform.render(ctx, &sources, &node.output)
+    }
+}
+
+/// Returns iterator that guarantees that nodes will be visited in
+/// the order that ensures that inputs of the current node were
+/// already visited in previous iterations
+fn graph_iter(scene: &Scene) -> impl Iterator<Item = Arc<Node>> {
+    let mut queue = vec![];
+    for (output_node, _output_texture) in scene.outputs.values() {
+        enqueue_nodes(output_node, &mut queue);
+    }
+    queue.into_iter()
+}
+
+fn enqueue_nodes(parent_node: &Arc<Node>, queue: &mut Vec<Arc<Node>>) {
+    let already_in_queue = queue.iter().any(|i| Arc::ptr_eq(parent_node, i));
+    if already_in_queue {
+        return;
+    }
+    for child in parent_node.inputs.iter() {
+        enqueue_nodes(child, queue);
+    }
+    queue.push(parent_node.clone());
+}
```

Comment:
I'm not really sure I understand, are you saying to move enque_nodes and put it as the closure inside graph_iter?
I would prefer to avoid it, alling closure recursively would be hacky (I would need to pass closure itself as an argument to it)

If you are saying to convert a recursive implementation to an iterative one and put it in grap_iter, then yes it would make sense, but it's a bit premature optimization in my opinion.

Reply:
you can put a regular function definition inside the scope of another function. this makes it visible only in that scope.

```rust
fn foo() {
    fn bar() -> i32 {
        return 42
    }
    
    println!("bar: {}", bar());
}
```

That's what I mean :smile:

Reply:
oh, I didn't know that syntax

I don't really like that pattern, If I needed sth like this I would probably prefer to use module instead. In this case render_loop modules is already quite small and enque_nodes is not accessible from outside that file, so I feel it's fine this way

Reply:
sure!